core/src/main/java/ai/timefold/solver/core/impl/move/streams/maybeapi/generic/provider/SwapMoveProvider.java

```diff
@@ -0,0 +1,95 @@
+import ai.timefold.solver.core.preview.api.domain.metamodel.VariableMetaModel;
+import ai.timefold.solver.core.preview.api.move.SolutionView;
+
+import org.jspecify.annotations.NullMarked;
+
+@NullMarked
+public class SwapMoveProvider<Solution_, Entity_>
+        implements MoveProvider<Solution_> {
+
+    private final PlanningEntityMetaModel<Solution_, Entity_> entityMetaModel;
+    private final List<PlanningVariableMetaModel<Solution_, Entity_, Object>> variableMetaModelList;
+
+    @SuppressWarnings("unchecked")
+    public SwapMoveProvider(PlanningEntityMetaModel<Solution_, Entity_> entityMetaModel) {
+        this.entityMetaModel = Objects.requireNonNull(entityMetaModel);
+        this.variableMetaModelList = entityMetaModel.variables().stream()
+                .flatMap(v -> {
+                    if (v instanceof PlanningVariableMetaModel<Solution_, Entity_, ?> planningVariableMetaModel) {
+                        return Stream.of((PlanningVariableMetaModel<Solution_, Entity_, Object>) planningVariableMetaModel);
+                    }
+                    return Stream.empty();
+                })
+                .toList();
+        if (variableMetaModelList.isEmpty()) {
+            throw new IllegalArgumentException("The entityClass (%s) has no basic planning variables."
+                    .formatted(entityMetaModel.type().getCanonicalName()));
+        }
+    }
+
+    public SwapMoveProvider(List<PlanningVariableMetaModel<Solution_, Entity_, Object>> variableMetaModelList) {
+        this.variableMetaModelList = Objects.requireNonNull(variableMetaModelList);
+        var entityMetaModels = variableMetaModelList.stream()
+                .map(VariableMetaModel::entity)
+                .distinct()
+                .toList();
+        this.entityMetaModel = switch (entityMetaModels.size()) {
+            case 0 -> throw new IllegalArgumentException("The variableMetaModelList (%s) is empty."
+                    .formatted(variableMetaModelList));
+            case 1 -> entityMetaModels.get(0);
+            default -> throw new IllegalArgumentException(
+                    "The variableMetaModelList (%s) contains variables from multiple entity classes."
+                            .formatted(variableMetaModelList));
+        };
+    }
+
+    @Override
+    public MoveProducer<Solution_> apply(MoveStreamFactory<Solution_> moveStreamFactory) {
+        var entityType = entityMetaModel.type();
+        var dataStream = moveStreamFactory.enumerate(entityType, false)
+                .join(entityType,
+                        filtering((SolutionView<Solution_> solutionView, Entity_ leftEntity, Entity_ rightEntity) -> {
+                            if (leftEntity == rightEntity) {
+                                return false;
+                            }
+                            var change = false;
+                            for (var variableMetaModel : variableMetaModelList) {
```

Comment:
We need to check all variables before returning. If any of them fail to accept the swap, then the change must not be accepted. For example, given e1(v1, v2) and e2(v3, v4). Let's assume that swapping v1 with v3 is accepted, but swapping v2 with v4 is not.

Reply:
Well spotted. 

This is very much a corner case though - the vast majority of our users only have one basic variable, in which case the code worked. And of those who have multiple basic variables, an even smaller number of them are using entity ranges. But you're right of course, it has to work regardless.